Sming/gdb/gdbuart.cpp

```diff
@@ -0,0 +1,419 @@
+	m_nputs(static_cast<const char*>(data), length);
+#endif
+	return gdb_uart_write(data, length);
+}
+
+size_t ATTR_GDBEXTERNFN gdbSendChar(char c)
+{
+#if GDBSTUB_ENABLE_DEBUG >= 3
+	m_putc(c);
+#endif
+	return gdb_uart_write_char(c);
+}
+
+size_t ATTR_GDBEXTERNFN gdbWriteConsole(const char* data, size_t length)
+{
+	GdbPacket packet;
+	packet.writeChar('O');
+	packet.writeHexBlock(data, length);
+	return length;
+}
+
+size_t ATTR_GDBEXTERNFN gdbSendUserData()
+{
+#if GDBSTUB_ENABLE_UART2
+	auto txbuf = user_uart == nullptr ? nullptr : user_uart->tx_buffer;
+	if(txbuf == nullptr) {
+		return 0;
+	}
+
+	size_t avail;
+	void* data;
+	while((avail = txbuf->getReadData(data)) != 0) {
+		size_t charCount;
+		unsigned used = uart_txfifo_count(GDB_UART);
+		unsigned space = UART_TX_FIFO_SIZE - used - 1;
+		if(gdb_state.attached) {
+			// $Onn#CC is smallest packet, for a single character, but we want to avoid that as it's inefficient
+			if(used >= 8) {
+				break;
+			}
+
+			charCount = std::min((space - 3) / 2, avail);
+			gdbWriteConsole(static_cast<const char*>(data), charCount);
+		} else {
+			charCount = gdb_uart_write(data, std::min(space, avail));
+		}
+
+		userDataSending = true;
+
+		user_uart->tx_buffer->skipRead(charCount);
+		if(charCount != avail) {
+			break; // That's all for now
+		}
+	}
+
+	return avail;
+#else
+	// UART2 disabled, no user data to send
+	return 0;
+#endif
+}
+
+void ATTR_GDBEXTERNFN gdbFlushUserData()
+{
+#if GDBSTUB_ENABLE_UART2
+	while(gdbSendUserData() != 0) {
+		wdt_feed();
+		system_soft_wdt_feed();
+	}
+#endif
+}
+
+#if GDBSTUB_ENABLE_UART2
+
+static void sendUserDataTask(uint32_t)
+{
+	sendUserDataQueued = false;
+
+	if(gdbSendUserData() == 0) {
+#if GDBSTUB_ENABLE_SYSCALL
+		// When all data has been sent, see if there's a pending syscall to send
+		gdbstub_syscall_execute();
+#endif
+	}
+}
+
+__forceinline void queueSendUserData()
+{
+	if(!sendUserDataQueued) {
+		System.queueCallback(sendUserDataTask);
+		sendUserDataQueued = true;
+	}
+}
+
+/**
+ * @brief Notify callback for user uart, called from uart driver
+ */
+static void userUartNotify(uart_t* uart, uart_notify_code_t code)
+{
+	switch(code) {
+	case UART_NOTIFY_AFTER_OPEN:
+		user_uart = uart;
+		break;
+
+	case UART_NOTIFY_BEFORE_CLOSE:
+		user_uart = nullptr;
+		break;
+
+	case UART_NOTIFY_AFTER_WRITE: {
+		/*
+		 * Driver calls here from uart_write() after data has been written to tx buffer.
+		 * Mostly we rely on task callback to transfer data from buffer to serial port,
+		 * but we need to send some data directly if the buffer is full otherwise we'll
+		 * loop indefinitely if UART_OPT_TXWAIT is set.
+		 */
+		if(uart->tx_buffer->isFull()) {
+			gdbSendUserData();
+		} else {
+			queueSendUserData();
+		}
+		break;
+	}
+
+	case UART_NOTIFY_WAIT_TX: {
+		gdbFlushUserData();
+		break;
+	}
+
+	case UART_NOTIFY_BEFORE_READ:
+		break;
+	}
+}
+#endif
+
+static void IRAM_ATTR doCtrlBreak()
+{
+	if(break_requests != 0) {

```

Comment:
One side-effect that I am experiencing in Eclipse, and probably other debuggers where you cannot send twice Ctrl-C, is that it looks like the debugging session has paused in the block below.

Reply:
> One side-effect that I am experiencing in Eclipse, and probably other debuggers where you cannot send twice Ctrl-C, is that it looks like the debugging session has **paused in the block below.**

@slaff Sorry, I don't understand - is this something you'd like me to look at before 3.8.0?